megamek/src/megamek/server/ServerHelper.java

```diff
@@ -530,6 +531,108 @@ public static boolean detectMinefields(Game game, Entity entity, Coords coords,
         return minefieldDetected;
     }
     
+    /**
+     * Checks to see if any units can detected hidden units.
+     */
+    public static boolean detectHiddenUnits(Game game, Entity detector, Coords detectorCoords,
+            Vector<Report> vPhaseReport, Server server) {
+        // If hidden units aren't on, nothing to do
+        if (!game.getOptions().booleanOption(OptionsConstants.ADVANCED_HIDDEN_UNITS)) {
+            return false;
+        }
+        
+        // Units without a position won't be able to detect
+        // check for this before calculating BAP range, as that's expensive
+        if ((detector.getPosition() == null) ||
+                (detectorCoords == null)) {
+            return false;
+        }
+       
+        int probeRange = detector.getBAPRange();
+        
+        // if no probe, save ourselves a few loops
+        if (probeRange <= 0) {
+            return false;
+        }
+                
+        // Get all hidden units in probe range
+        List<Entity> hiddenUnits = new ArrayList<>();
+        for (Coords coords : detectorCoords.allAtDistance(probeRange)) {
+            for (Entity entity : game.getEntitiesVector(coords, true)) {
+                if (entity.isHidden() && entity.isEnemyOf(detector)) {
+                    hiddenUnits.add(entity);
+                }
+            }
+        }
```

Comment:
```suggestion
        List<Entity> hiddenUnits = detectorCoords.allAtDistance(probeRange).stream()
                .flatMap(coords -> game.getEntitiesVector(coords, true).stream()
                        .filter(entity -> entity.isHidden() && entity.isEnemyOf(detector)))
                .collect(Collectors.toList());
```

Reply:
Not sure that this makes it easier to read or debug or has any performance benefits.

Reply:
Readability is arguably better, it is far more modern, and it scales far better (given player proclivities this is key).

Reply:
Modern != better necessarily, and I'm not really sure how it scales better. Do the "stream" functions not have to loop through each element in the collection?

Reply:
This is discussed in the Stream JavaDoc, and can be better explained by either reading that or searching google for why to use streams over loops.

https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html

Reply:
I read the document, and while it promises "greater parallelism", that does not happen in reality (I figured it was the case because similar technology in the .NET environment is also slower than just basic loops). I ran some tests on a 100-something x 100-something map, with a BAP-equipped Loki on the "detector" end and ~40 hidden Lokis. I set the BAP range to 100 artificially to get meaningful numbers. The raw data indicates that the "collection" method you propose is actually slower by about 2ms per run of the code in question, and much worse if I discard the high outliers.

So, I will not be implementing your suggestion. 

Here's the raw data, including a save. The logging is "entityID: time in ms to run detect hidden unit routine".

[collectdata.txt](https://github.com/MegaMek/megamek/files/7599899/collectdata.txt)
[loopdata.txt](https://github.com/MegaMek/megamek/files/7599900/loopdata.txt)
[quicksave.sav.gz](https://github.com/MegaMek/megamek/files/7599902/quicksave.sav.gz)

The code I used for time measurement is as follows (comment/uncomment the relevant implementation you want to test).

// Get all hidden units in probe range
        long currentMillis = System.currentTimeMillis();
        
        List<Entity> hiddenUnits = new ArrayList<>();
        for (Coords coords : detectorCoords.allAtDistanceOrLess(probeRange)) {
            for (Entity entity : game.getEntitiesVector(coords, true)) {
                if (entity.isHidden() && entity.isEnemyOf(detector)) {
                    hiddenUnits.add(entity);
                }
            }
        }
        
        /*List<Entity> hiddenUnits = detectorCoords.allAtDistanceOrLess(probeRange).stream()
                .flatMap(coords -> game.getEntitiesVector(coords, true).stream()
                        .filter(entity -> entity.isHidden() && entity.isEnemyOf(detector)))
                .collect(Collectors.toList());*/
        
        long endMillis = System.currentTimeMillis();
        long result = endMillis - currentMillis;
        MegaMek.getLogger().error(String.format("%d: %d", detector.getId(), result));

Reply:
However, this did have the side effect of uncovering a bug in my implemetation, which is that I was only checking the outer edge of the BAP radius and nothing inside it.

Reply:
That's interesting, and... I agree on the performance bit now. I do think it is more readable, but that's (especially in this case) very much a personal viewpoint.

Sorry about the delayed response, missed the reply messages on Discord.